src/Sensors/IMU/IMU.cpp

```diff
@@ -54,14 +66,98 @@ namespace mmfs
+        Quaternion q_w_BI_dot = w_b * q_w_BI * -.5;
+
+        // 2. Propogate gyro orientation
+        q_w_BI = q_w_BI + q_w_BI_dot * dt;
+        q_w_BI.normalize();
+        //-------------------------------------------------//
+
+        //----------------- ACCEL ORIENTATION --------------//
+        // 3. Find the predicted gravity vector in the inertial frame
+        Quaternion a_b = Quaternion{0, accelerationVec};
+        if(!(a_b.magnitude() > 0)){
+            // Accel has no roll or pitch contribution to be made
+            orientation = q_w_BI.conjugate();
+            return;
+        }
+        a_b.normalize();
+        Quaternion g_p_I = q_w_BI.conjugate() * a_b * q_w_BI;
+        // 4. Determine delta acc orientation
+        if(g_p_I.z() == -1){
+            // TODO this shouldn't happen though
+            // This prob needs some tolerance too
+            
+            // Temp solution (I just made this up lol)
+            g_p_I.z() = -.9;
+            g_p_I.w() = .01;
+            g_p_I.normalize();
+        }
+
+        Quaternion delta_q_acc = Quaternion{sqrt((g_p_I.z()+1)/2), -g_p_I.y()/sqrt(2*(g_p_I.z()+1)), g_p_I.x()/sqrt(2*(g_p_I.z()+1)), 0};
+        
+        // 5. Adaptive gain interpolation to reduce high freq noise
+        double alpha = adaptiveAccelGain(accel_best_filtering_at_static, .1, .2);
+        delta_q_acc = delta_q_acc.interpolation(Quaternion{1,0,0,0}, alpha, .9);
+
+        // 6. Combine with gryo orientation to correct roll and pitch
+        Quaternion q_wa_BI = q_w_BI * delta_q_acc;
+        q_wa_BI.normalize();
+        //-------------------------------------------------//
+
+        //----------------- MAG ORIENTATION --------------//
+
+        // 7. Get magnetic field vector in interial frame
+        Quaternion m_B = Quaternion{0, magField};
+        if(!(m_B.magnitude() > 0)){
+            // Mag Field has no yaw contribution to be made
+            orientation = q_wa_BI.conjugate();
+            return;
+        }
+        m_B.normalize();
+        Quaternion m_I = q_wa_BI.conjugate() * m_B * q_wa_BI;
+
+        // 8. Determine delta mag orientation
+        double Gamma = m_I.x()*m_I.x() + m_I.y()*m_I.y();
+
+        Quaternion delta_q_mag = Quaternion{sqrt(Gamma + (m_I.x()*sqrt(Gamma)))/sqrt(2*Gamma), 0, 0, m_I.y()/sqrt(2*(Gamma+ (m_I.x()*sqrt(Gamma))))};
+
+        // 9. Interpolation to reduce high freq noise 
+        delta_q_mag = delta_q_mag.interpolation(Quaternion{1,0,0,0}, mag_best_filtering_at_static, .9);
+
+        // 10. Combine with gryo/acc orientation to correct yaw
+        Quaternion q_BI = q_wa_BI * delta_q_mag;
+        q_BI.normalize();
+        //-------------------------------------------------//
+
+        orientation = q_BI.conjugate();
     }
 
-    bool IMU::begin(bool useBiasCorrection)
+    double IMU::adaptiveAccelGain(double alphaBar, double t_1, double t_2)
```

Comment:
What calls this function?

Reply:
This function can be thought of as util functions native to IMUs and get called during the hardware read() step if that hardware needs it to find orientation.

For this specifically it is called by quaternionBasedComplimentaryFilter as well